OpenRA.Game/Activities/Activity.cs

```diff
@@ -70,12 +71,18 @@ public Activity TickOuter(Actor self)
 			if (State == ActivityState.Done)
 				throw new InvalidOperationException("Actor {0} attempted to tick activity {1} after it had already completed.".F(self, GetType()));
 
-			if (State == ActivityState.Queued)
+			// ActivityState.Queued and firstRun are not equivalent.
+			// A queued activity that has been canceled will have ActivityState.Canceling,
+			// and we must still call OnFirstRun before Tick/OnLastRun.
```

Comment:
I've been out of the loop for a bit, so I don't quite understand where the need to run a cancelled activity at all comes from. That sounds horribly broken. In the original implementation the getters would have made sure that cancelled activities weren't even visible on the queue, but I see that they've been replaced by default getters.

Reply:
My initial approach here was to have Cancel change `ActivityState.Queued` directly to `ActivityState.Done`, and the `return NextActivity` below first check whether it is `Queued` vs `Done`.

I had two concerns that made me change to the current approach:
* Its not as simple as just returning `null` instead of `NextActivity` - a non-queued activity may be queued after this.
* The current activity behaviour has had a lot of testing, and i'm worried that changing this behaviour may cause unexpected regressions.

If this was aimed for Next + 1 I would probably have stuck with it, but I wanted to be more conservative for the hotfix.

Reply:
> My initial approach here was to have Cancel change ActivityState.Queued directly to ActivityState.Done

Wouldn't that make the `Cancelling` state unnecessary?

> Its not as simple as just returning null instead of NextActivity - a non-queued activity may be queued after this

(You mean non-cancelled, right?)
Yes, it should be that simple*. If the getter returned null for cancelled activities, this situation could never arise. That was one of the two reasons why I did it that way - the other being to prevent cancelled activities from being run on the queue.

And if you really want to selectively only cancel the *next* activity, but not the ones following it, then that next activity should really be a child.

But I may be missing something. My impression is that OpenRA's code has become even more convoluted in the past few years, so it's likely that I don't know about edge case that should be considered.

*Unless the cancelled activity's constructor has side-effects. Which would be bug.

Reply:
> Wouldn't that make the `Cancelling` state unnecessary?

The `Canceling` state is still needed for activities in the `Active` state, which may require an unknown number of ticks to get themselves back into a safe configuration (e.g. moving back into the cell grid, (un)deploying, (un)docking) before completing.

> And if you really want to selectively only cancel the *next* activity, but not the ones following it, then that next activity should really be a child.

The issue comes when we want to cancel *all* activities, but have a specific reason to not cancel the current one yet (e.g. here, where we would lose target lines, or [TransformsIntoMobile](https://github.com/OpenRA/OpenRA/blob/308c64c7b1980f05bf04f98425fb8fea92d33608/OpenRA.Mods.Common/Traits/Buildings/TransformsIntoMobile.cs#L110-L115) where the undeploy needs to continue but we replace the move order(s) issued to the new actor).

Otherwise I agree with your argument here, and I don't mind changing this back if the others are happy to take the increased risk of regression.

Reply:
88314f8119a6bf1bae651186d91ccfd27515bc6c shows a refined version of this other approach.

Reply:
8c3cadb651 reworks the comment following discussion in IRC.

Reply:
Rebased onto the updated commit.